### PR TITLE
add opcodes: PACKSTRUCT, PACKMAP

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo-vm/issues/442

### CODE REVIEW

**Comment by shargon** [comment]
> What is the benefit of FILL for an Array over PACK? Seems like the latter would still be cheaper and faster

No benefict, just different behaviours

> How will we determine the price for FILL? My assumption is that filling an Array is computationally cheaper than filling a Map

It's the same problem for `PACK` and `PACKMAP`

**Comment by ixje** [comment]
> > How will we determine the price for FILL? My assumption is that filling an Array is computationally cheaper than filling a Map
> 
> It's the same problem for `PACK` and `PACKMAP`

How? `PACK` and `PACKMAP` are 2 instructions, therefor each can have their own price. `FILL` is 1 instruction covering 2 use-cases (ARRAY & MAP)

**Comment by shargon** [comment]
> How? PACK and PACKMAP are 2 instructions, therefor each can have their own price. FILL is 1 instruction covering 2 use-cases (ARRAY & MAP)

I mean, `PACK` one item cost the same as `PACK` 100

**Comment by ixje** [comment]
> > How? PACK and PACKMAP are 2 instructions, therefor each can have their own price. FILL is 1 instruction covering 2 use-cases (ARRAY & MAP)
> 
> I mean, `PACK` one item cost the same as `PACK` 100

Aaaaah good point, I didn't noticed that. I was thinking of popping single stack items and appending for `ARRAY` vs popping key/value pairs for `MAP`.

**Comment by Qiao-Jin** [comment]
Maybe a inverse opcode (Unload?) is needed?

**Comment by erikzhang** [comment]
I prefer to add `PACKSTRUCT` and `PACKMAP`.

**Comment by erikzhang** [comment]
@shargon `UNPACK` needs to support map.

**Review by erikzhang — DISMISSED** [review]

File: src/neo-vm/ExecutionEngine.cs [hunk 1]
```diff
@@ -1004,10 +1033,26 @@ private void ExecuteInstruction()
...
+                            case VMArray array:
+                                for (int i = array.Count - 1; i >= 0; i--)
+                                {
+                                    Push(array[i]);
+                                }
+                                break;
+                            default:
+                                throw new InvalidOperationException($"Invalid type for {instruction.OpCode}: {compound.Type}");
+                        }
+                        Push(compound.Count);
```
Comment (shargon) [inline src/neo-vm/ExecutionEngine.cs:1055]: Is not *2 when map?

Comment (erikzhang) [reply src/neo-vm/ExecutionEngine.cs:1055]: No.

**Review by erikzhang — APPROVED** [review]

**Comment by ixje** [comment]
@erikzhang what will be the opcode prices for `PACKMAP` and `PACKSTRUCT`?

**Comment by erikzhang** [comment]
The same as `PACK`.

**Comment by ixje** [comment]
Thanks!

**Pull request merged**